### PR TITLE
[Symex] Dereference function call

### PR DESCRIPTION
**Pull request by lucasccordeiro**
This PR allows dereferencing function pointer for empty parameters. In particular, C allows function calls as `func()`, but C++ not.
SV-COMP is full of examples that contains function calls for the `pthread_create` function as `void* func()` instead of `void* func(void *arg)`.

### CODE REVIEW

File: src/goto-symex/symex_function.cpp [hunk 1]
```diff
@@ -405,8 +405,18 @@ void goto_symext::symex_function_call_deref(const expr2tc &expr)
...
-      return true;
+    {
+      // C allows function calls as func()
+      // This happens in some multi-threaded C programs from SV-COMP
+      // i.e. void* func() instead of void* func(void *arg)
+      if(
+        !has_prefix(to_symbol2t(sym).thename.as_string(), "c:@") &&
+        !ct.arguments.size())
+        return true;
+    }
```
Comment (mikhailramalho) [inline src/goto-symex/symex_function.cpp]: Do we really need to be so conservative this far down into the verification? 

What if we simply check if there are enough arguments to call the function?

Comment (lucasccordeiro) [reply src/goto-symex/symex_function.cpp]: Fixed. I agree with you.

Our C++ clang frontend already detects the illegal cases. So, we don't need to be so conservative here.

**Comment by lucasccordeiro** [comment]
After fixing the dereference function call issue, this PR has enabled 6 regression tests, but disabled 3 regression tests. I have open another ticket to investigate this problem (https://github.com/esbmc/esbmc/issues/583), which also happens in the SV-COMP benchmarks.

**Review by fbrausse — COMMENTED** [review]
C defines a few different ways function declarations can look like:
```c
int a(); // K&R-style, no function prototype, promoted types should be compatible with definition, could be any #args
int b(void); // zero arguments allowed
int c(int, int, int); // same as b, but with 3 arguments
int d(int, int, int, ...); // different from c, >= 3 argument(s)
```
At the moment, the clang frontend does not differentiate between `a` and `b` (it would need to check whether `fd.getType()->getAs<clang::FunctionProtoType>()` is null in `clang_c_convertert::get_function()`) -- both have the same representation in terms of their types. For some reason though, it marks (1) function calls `int x = f()`: https://github.com/esbmc/esbmc/blob/228f3fca7222ad7ca376e62485f60934c8b1cae4/src/clang-c-frontend/clang_c_convert.cpp#L853-L855 and (2) `a` *and* `b` as variadic:
https://github.com/esbmc/esbmc/blob/228f3fca7222ad7ca376e62485f60934c8b1cae4/src/clang-c-frontend/clang_c_convert.cpp#L564-L566 From what I can see, ESBMC does not consider the notion of "prototype" in types of symbols. `a` gets overwritten if it is given a prototype in some translation unit (either via non-K&R-style declaration or via definition). So the prototype-less `a` (empty parameter list, variadic) only happens when no function body is available and is given the same representation as `b`.

Depending on the reason behind considering particular symbols to be "illegal targets", it might be OK to consider `a` a "legal target" for any amount of operands in the call -- can't execute it in that case anyway; but I would consider calling `b` with >= 1 arguments undefined (and therefore rejected by ESBMC; could be part of #580, not necessarily this PR). These changes look reasonable, but should probably also take `code_type2t::ellipsis` into account in order to keep `d` "legal" when called with 4 parameters.

**Comment by fbrausse** [comment]
PS: Fortunately, the frontend saves us from the cases
```c
int e(...); // rejected by clang
b(5); // rejected by clang
```

**Comment by fbrausse** [comment]
PPS: A K&R-style definition `int f() { return 42; }` technically still has no prototype (until C2x, see [n2432](http://www.open-std.org/jtc1/sc22/wg14/www/docs/n2432.pdf)). Clang warns, GCC just accepts calling it with any number of arguments.

File: src/goto-symex/symex_function.cpp [hunk 2]
```diff
@@ -405,7 +405,12 @@ void goto_symext::symex_function_call_deref(const expr2tc &expr)
       return true;
 
     const code_type2t &ct = to_code_type(sym->type);
-    if(ct.arguments.size() != call.operands.size())
+
+    // We check whether the number of arguments matches.
+    // C allows function calls as func(), i.e.,
+    // void* func() instead of void* func(void *arg).
+    // Our C++ front-end detects this as a parsing error.
+    if(ct.arguments.size() && ct.arguments.size() != call.operands.size())
```
Comment (fbrausse) [inline src/goto-symex/symex_function.cpp:415]: As described, `int a();` is K&R-style declaration that lacks a prototype. It is unknown how such things could be called without a definition. If there is a definition, the call must match its parameter list. That is, calling `f(5)` is undefined (even though implementations seem to allow it). Existence of a prototype depends on the declaration, not the call. C++ does not allow declarations w/o prototypes. The comment should talk about the declaration.

What about the variadic case (e.g., `d`)?

Comment (lucasccordeiro) [reply src/goto-symex/symex_function.cpp:415]: I'm unsure whether we support the variadic case.

@mikhailramalho and @rafaelsamenezes: do you have any thoughts here?

**Comment by fbrausse** [comment]
> SV-COMP is full of examples that contains function calls for the pthread_create function as void* func() instead of void* func(void *arg)

They are all wrong :) Does SV-COMP specify the C dialect they want to adhere to? C2x is going to make `a()` equivalent to `b(void)`. SV-COMP will have to adjust...

**Comment by fbrausse** [comment]
Back to reality:
```c
#include <assert.h>
void g(int x, int y) { assert(0); }
void trampo(void (*g)()) { g(5); }
int main() { trampo(g); }
```
violates the C std in a similar way by not giving enough arguments to `g`, but is accepted by implementations. ESBMC removes the call to `g` in the `illegal_filter` and sees no problem with this program. Should the filter be more lenient?

**Review by fbrausse — APPROVED** [review]
I'm OK with moving the UB checks to #580, variadics to #584 and other implementation-accepted but uncommon behaviour to #585. LGTM.

**Comment by lucasccordeiro** [comment]
I have created two issues to deal with @fbrausse's comments in other PRs: https://github.com/esbmc/esbmc/issues/584 and https://github.com/esbmc/esbmc/issues/585.

File: src/goto-symex/symex_function.cpp (hunk 2 above)
Comment (rafaelsamenezes) [reply src/goto-symex/symex_function.cpp:415]: > I'm unsure whether we support the variadic case.

AFAIK, variadic functions are not supported in ESBMC. When we eventually support variadic functions we can just add a member `is_variadic()` and check it here.

**Review by rafaelsamenezes — APPROVED** [review]

File: src/goto-symex/symex_function.cpp (hunk 2 above)
Comment (mikhailramalho) [reply src/goto-symex/symex_function.cpp:415]: We do support variadic functions, but only on 32 bits. You can find how it's modeled in the builtin_functions file in the goto dir, and how we try to handle it when creating a function frame during symex). I'm not sure how we handle variadic functions through function pointers. Is that even possible?

My suggestion here is to check if we have enough arguments to fill all the function parameters, cast the arguments to the type of the parameter (maybe this should done only on C, given that C++ actually cares about argument/parameters type mismatch), and call the function.

Comment (fbrausse) [reply src/goto-symex/symex_function.cpp:415]: > variadic functions through function pointers. Is that even possible?

I believe it is allowed. Syntactically it's just what you'd expect: `int (*p)(const char *, ...) = printf` and works in practice. Note that `int (*p)() = printf` is the not same (C2x or not): the calling conventions differ.

**Pull request merged**